tests/unit/lifecycle/LifecycleBucketProcessor.spec.js

```diff
@@ -197,4 +205,73 @@ describe('Lifecycle Bucket Processor', () => {
+
+    describe('_processBucketEntry', () => {
+
+        [
+            {
+                it: 'should use the forceLegacyListing when listing version is not ' +
+                    'specified in the Kafka messages (default case)',
+                forceLegacyListing: null,
+                kafkaMessage: bucketProcessorEntry,
+                validateTask: task => task instanceof LifecycleTaskV2,
+            },
+            {
+                it: 'should use the forceLegacyListing when listing version is not ' +
+                    'specified in the Kafka messages (v1 case)',
+                forceLegacyListing: true,
+                kafkaMessage: bucketProcessorEntry,
+                validateTask: task => task instanceof LifecycleTask && !(task instanceof LifecycleTaskV2),
+            },
+            {
+                it: 'should use the forceLegacyListing when listing version is not ' +
+                    'specified in the Kafka messages (v2 case)',
+                forceLegacyListing: false,
+                kafkaMessage: bucketProcessorEntry,
+                validateTask: task => task instanceof LifecycleTaskV2,
+            },
+            {
+                it: 'should use the listing type specified in the kafka message (v1 case)',
+                forceLegacyListing: null,
+                kafkaMessage: bucketProcessorV1Entry,
+                validateTask: task => task instanceof LifecycleTask && !(task instanceof LifecycleTaskV2),
+            },
+            {
+                it: 'should use the listing type specified in the kafka message (v2 case)',
+                forceLegacyListing: false,
+                kafkaMessage: bucketProcessorV2Entry,
+                validateTask: task => task instanceof LifecycleTaskV2,
+            },
+        ].forEach(opts => {
+            it(opts.it, () => {
+                if (opts.forceLegacyListing !== null) {
+                    lbp._lcConfig.forceLegacyListing = opts.forceLegacyListing;
+                }
+
+                lbp.clientManager = {
+                    getS3Client: () => sinon.stub(),
+                    getBackbeatMetadataProxy: () => sinon.stub(),
+                };
+
+                sinon.stub(lbp, '_getBucketLifecycleConfiguration').yields(null, {
+                    Rules: {
+                        Status: 'Enabled',
+                        Transitions: [{
+                            Days: 10,
+                            StorageClass: 'azure'
+                        }],
+                        ID: 'dac36d89-0005-4c78-8e00-7e9ace06a9c4'
+                    }
+                });
+
+                lbp._internalTaskScheduler = [];
+
+                lbp._processBucketEntry(opts.kafkaMessage, err => {
+                    assert.ifError(err);
+                    assert.strictEqual(lbp._internalTaskScheduler.length, 1);
+                    assert(opts.validateTask(lbp._internalTaskScheduler[0].task));
+                });
+            });
+        });
```

Comment:
This is not testing the function at all, as we do the checks in the callback

```suggestion
            it(opts.it, done => {
                if (opts.forceLegacyListing !== null) {
                    lbp._lcConfig.forceLegacyListing = opts.forceLegacyListing;
                }

                lbp.clientManager = {
                    getS3Client: () => sinon.stub(),
                    getBackbeatMetadataProxy: () => sinon.stub(),
                };

                sinon.stub(lbp, '_getBucketLifecycleConfiguration').yields(null, {
                    Rules: {
                        Status: 'Enabled',
                        Transitions: [{
                            Days: 10,
                            StorageClass: 'azure'
                        }],
                        ID: 'dac36d89-0005-4c78-8e00-7e9ace06a9c4'
                    }
                });

                lbp._internalTaskScheduler = [];

                lbp._processBucketEntry(opts.kafkaMessage, err => {
                    if (err) {
                        return done(err);
                    }
                    assert.strictEqual(lbp._internalTaskScheduler.length, 1);
                    assert(opts.validateTask(lbp._internalTaskScheduler[0].task));
                    done();
                });
            });
        });
```

note: it also seems like the stub is not working for some reason...

Reply:
oh yeah good catch !